python/tvm/relay/transform/quantization/quantized_operators/dense.py

```diff
@@ -0,0 +1,183 @@
+        * data_zero_point
+        * weight_zero_point
+    )
+
+    # TODO: simulate overflow for other data types
+
+    # return first_term, second_term, third_term, fourth_term
+    return (first_term - second_term - third_term + fourth_term), QParams(
+        data_scale * weight_scale, 0, simulated_accumulation_dtype
+    )
+
+
+def generate_static_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    return generate_generic_quantized_dense(
+        data,
+        weight,
+        data_qparams,
+        weight_qparams,
+        internal_accumulation_dtype=accumulation_dtype,
+        simulated_accumulation_dtype=accumulation_dtype,
+        out_units=out_units,
+        in_units=in_units,
+    )
+
+
+def generate_simulated_quantized_dense(
+    data: tvm.relay.Expr,
+    weight: tvm.relay.Expr,
+    data_qparams: QParams,
+    weight_qparams: QParams,
+    simulated_accumulation_dtype: str = "int32",
+    out_units: Optional[int] = None,
+    in_units: Optional[int] = None,
+) -> Tuple[tvm.relay.Expr, QParams]:
+    return generate_generic_quantized_dense(
+        data,
+        weight,
+        data_qparams,
+        weight_qparams,
+        internal_accumulation_dtype="float32",
+        simulated_accumulation_dtype=simulated_accumulation_dtype,
+        out_units=out_units,
+        in_units=in_units,
+    )
+
+
+def example_dense_simulated(n, in_units, out_units, seed=42):
+    np.random.seed(seed=seed)
+    data_arr = np.random.randint(-10, 10, size=(n, in_units)).astype("float32")
+    weight_arr = np.random.randint(-10, 10, size=(out_units, in_units)).astype("float32")
+
+    var_creator = AffineQuantizationVarCreator()
+    data = relay.Var("data")
+    weight = relay.Var("weight")
+    data_qparams = var_creator.get_qparams("dense_data")
+    weight_qparams = var_creator.get_qparams("dense_weight")
+    dense_output, output_qparams = generate_simulated_quantized_dense(
+        data, weight, data_qparams, weight_qparams, in_units=in_units, out_units=out_units
+    )
+    f = relay.Function(
+        [data, weight, data_qparams.zero_point, weight_qparams.zero_point], dense_output
+    )
+    mod = tvm.ir.IRModule.from_expr(f)
+    intrp = relay.create_executor(kind="debug", mod=mod)
+    return intrp.evaluate(f)(data_arr, weight_arr, 0.0, 0.0).asnumpy()
+
+
+def example_dense_static_quantized(n, in_units, out_units, seed=42):
+    np.random.seed(seed=seed)
+    data_arr = np.random.randint(-10, 10, size=(n, in_units)).astype("int8")
+    weight_arr = np.random.randint(-10, 10, size=(out_units, in_units)).astype("int8")
+
+    var_creator = AffineQuantizationVarCreator()
+    data = relay.Var("data")
+    weight = relay.Var("weight")
```

Comment:
Are you assuming that the `data` and `weight` vars being passed in are already quantized using the scales and zero points?

Reply:
For now, I'm going to change it so it assumes everything is fp32.

The invariant I am assuming for these "translation" units is fp32 in, fp32 out.